lib/rules/no-shadow-restricted-names.js

```diff
@@ -4,6 +4,19 @@
  */
 "use strict";
 
+/**
+ * Determines if a variable safely shadows undefined.
+ * This is the case when a variable named `undefined` is never assigned to a value (i.e. it always shares the same value
+ * as the global).
+ * @param {eslintScope.Variable} variable The variable to check
+ * @returns {boolean} true if this variable safely shadows `undefined`
+ */
+function safelyShadowsUndefined(variable) {
+    return variable.name === "undefined" &&
+        variable.references.every(ref => !ref.isWrite()) &&
+        variable.defs.every(def => def.node.type === "VariableDeclarator" && def.node.init === null);
+}
```

Comment:
what if `var undefined = void 0;`-- it was always safe.

Reply:
Have you seen anyone use this pattern? I'm fine with adding it if it's used, but it's not possible for us to handle every case that always evaluates to `undefined`, so I'm not sure where we should draw the line.

Reply:
No, it just comes to my mind when reviewing the code 😄. 
or we can merge as is, and add this case if having some reports.